include/util/d_ary_heap.hpp

```diff
@@ -0,0 +1,113 @@
+#pragma once
+
+#include <boost/assert.hpp>
+#include <functional>
+#include <limits>
+#include <utility>
+#include <vector>
+
+namespace osrm::util
+{
+template <typename HeapData, int Arity, typename Comparator = std::less<HeapData>> class DAryHeap
```

Comment:
Having the reorder handler as a template parameter seems somewhat confusing. 

Do have a need to specify multiple ones at all? If this is needed, should the be implemented inside the heap?

Can we possibly aim for a simpler implementation of this?

Reply:
>Having the reorder handler as a template parameter seems somewhat confusing.

Well, the intention here was to be able to pass lambda and avoid using std::function which has certain performance impact... Such template parameters is frequent pattern for such things - if you want to pass lambda, compiler will infer type of this lambda for you in a most efficient way. At the same time I agree that probably having the fact that we pass only single type of function in there worth trying to make this more obvious (not sure it will be simple though, but I'll try 😄 )

>If this is needed, should the be implemented inside the heap?

What do you mean by "inside the heap"? In my initial implementation `DAryHeap` wasn't a separate class and all code from here "lived" directly in `QueryHeap`, but I found it a bit messy and decided to decouple things a bit to keep code structure a bit similar to what we have now (QueryHeap and separate boost::d_ary_heap before and QueryHeap and separate DAryHeap after). I could try to return back to initial implementation...

WDYT?

Reply:
Or you mean that implementation of `reorderHandler` should be a part of DAryHeap? I am not sure it is also doable without merging DAryHeap to QueryHeap - this `reorderHandler` is kind of a "glue" between `DAryHeap` and `QueryHeap`...

Reply:
Another option I see is... make it super straightforward. DAryHeap now is kind of generic DAryHeap implementation, but we could "couple" it to our needs, i.e. just take reference to `inserted_nodes` from QueryHeap and do all things we are currently doing in reorderHandler directly on that vector. Should also work without issues as well.

Anyway @DennisOSRM would love to hear your opinion on this please 😄

Reply:
SO question about passing lambdas as parameters https://stackoverflow.com/questions/6458612/proper-way-to-receive-a-lambda-as-parameter-by-reference

Reply:
And one more blogpost with concrete example of why template can be more preferable here… https://wolchok.org/posts/cxx-trap-2-std-function/

Reply:
@DennisOSRM WDYT? Tbh I am not sure I have good ideas how to remove template parameter here - all ideas have their own drawbacks. May be you have some? Or may be worth introducing a bit of complexity in favor of performance? 😀